_In this week, we will use basketball data downloaded from NBA.com to demonstrate how to import data into Python, how to clean up data before conducting any data analyses, as well how to describe and summarize data._ 


## Importing data into Python
### Before we import the dataset into Jupyter notebook, we need to first import the Python libraries that we will use to analyze the data
- pandas
- numpy



In [1]:
import pandas as pd
import numpy as np

#### In our data repository, we have a dataset that contains NBA team information. Let's import this dataset into the Jupyter notebook.

In [2]:
NBA_Teams=pd.read_csv("../../Data/Week 2/nba_teams.csv")

We can take a quick look at the data we imported by displaying the dataset. 

In [3]:
display(NBA_Teams)

Unnamed: 0 ABBREVIATION           CITY               FULL_NAME  \
0            0          ATL        Atlanta           Atlanta Hawks   
1            1          BOS         Boston          Boston Celtics   
2            2          CLE      Cleveland     Cleveland Cavaliers   
3            3          NOP    New Orleans    New Orleans Pelicans   
4            4          CHI        Chicago           Chicago Bulls   
5            5          DAL         Dallas        Dallas Mavericks   
6            6          DEN         Denver          Denver Nuggets   
7            7          GSW   Golden State   Golden State Warriors   
8            8          HOU        Houston         Houston Rockets   
9            9          LAC    Los Angeles    Los Angeles Clippers   
10          10          LAL    Los Angeles      Los Angeles Lakers   
11          11          MIA          Miami              Miami Heat   
12          12          MIL      Milwaukee         Milwaukee Bucks   
13          13          MIN      Minnesota  Minnesota Timberwolves   
14          14          BKN       Brooklyn           Brooklyn Nets   
15          15          NYK       New York         New York Knicks   
16          16          ORL        Orlando           Orlando Magic   
17          17          IND        Indiana          Indiana Pacers   
18          18          PHI   Philadelphia      Philadelphia 76ers   
19          19          PHX        Phoenix            Phoenix Suns   
20          20          POR       Portland  Portland Trail Blazers   
21          21          SAC     Sacramento        Sacramento Kings   
22          22          SAS    San Antonio       San Antonio Spurs   
23          23          OKC  Oklahoma City   Oklahoma City Thunder   
24          24          TOR        Toronto         Toronto Raptors   
25          25          UTA           Utah               Utah Jazz   
26          26          MEM        Memphis       Memphis Grizzlies   
27          27          WAS     Washington      Washington Wizards   
28          28          DET        Detroit         Detroit Pistons   
29          29          CHA      Charlotte       Charlotte Hornets   

            ID       NICKNAME                 STATE  YEAR_FOUNDED  
0   1610612737          Hawks               Atlanta          1949  
1   1610612738        Celtics         Massachusetts          1946  
2   1610612739      Cavaliers                  Ohio          1970  
3   1610612740       Pelicans             Louisiana          2002  
4   1610612741          Bulls              Illinois          1966  
5   1610612742      Mavericks                 Texas          1980  
6   1610612743        Nuggets              Colorado          1976  
7   1610612744       Warriors            California          1946  
8   1610612745        Rockets                 Texas          1967  
9   1610612746       Clippers            California          1970  
10  1610612747         Lakers            California          1948  
11  1610612748           Heat               Florida          1988  
12  1610612749          Bucks             Wisconsin          1968  
13  1610612750   Timberwolves             Minnesota          1989  
14  1610612751           Nets              New York          1976  
15  1610612752         Knicks              New York          1946  
16  1610612753          Magic               Florida          1989  
17  1610612754         Pacers               Indiana          1976  
18  1610612755          76ers          Pennsylvania          1949  
19  1610612756           Suns               Arizona          1968  
20  1610612757  Trail Blazers                Oregon          1970  
21  1610612758          Kings            California          1948  
22  1610612759          Spurs                 Texas          1976  
23  1610612760        Thunder              Oklahoma          1967  
24  1610612761        Raptors               Ontario          1995  
25  1610612762           Jazz                  Utah          1974

_This dataset provides some basic information of the NBA teams._ 

For a dataset, each row represents an observation, i.e., a team in this dataset and each column represents a variable which contains information of a characteristics of the observation. A variable can take different values in different situations. The number of observation in a dataset represents the size of our sample and the number of variables represents the richness of information in our dataset. 

#### We can use the “shape” function in Python to see how many variables and observations in our dataset.




In [4]:
NBA_Teams.shape

(30, 8)

_We can see that there are 30 observations (rows) and 8 variables (columns)._

### Renaming Variables
####  We can rename a variable using the “rename” function in Python.
_Inplace parameter_
- True: replace the old variable with a new name; 
- False: create a new variable with the new name.

The first variable is unnamed, let's rename it to be "TEAM_NUMBER"; let's also rename "ID" to "TEAM_ID." 

In [7]:
NBA_Teams.rename(columns = {'Unnamed: 0':'TEAM_NUMBER', 'ID':'TEAM_ID'}, inplace = True)
display(NBA_Teams)

TEAM_NUMBER ABBREVIATION           CITY               FULL_NAME  \
0             0          ATL        Atlanta           Atlanta Hawks   
1             1          BOS         Boston          Boston Celtics   
2             2          CLE      Cleveland     Cleveland Cavaliers   
3             3          NOP    New Orleans    New Orleans Pelicans   
4             4          CHI        Chicago           Chicago Bulls   
5             5          DAL         Dallas        Dallas Mavericks   
6             6          DEN         Denver          Denver Nuggets   
7             7          GSW   Golden State   Golden State Warriors   
8             8          HOU        Houston         Houston Rockets   
9             9          LAC    Los Angeles    Los Angeles Clippers   
10           10          LAL    Los Angeles      Los Angeles Lakers   
11           11          MIA          Miami              Miami Heat   
12           12          MIL      Milwaukee         Milwaukee Bucks   
13           13          MIN      Minnesota  Minnesota Timberwolves   
14           14          BKN       Brooklyn           Brooklyn Nets   
15           15          NYK       New York         New York Knicks   
16           16          ORL        Orlando           Orlando Magic   
17           17          IND        Indiana          Indiana Pacers   
18           18          PHI   Philadelphia      Philadelphia 76ers   
19           19          PHX        Phoenix            Phoenix Suns   
20           20          POR       Portland  Portland Trail Blazers   
21           21          SAC     Sacramento        Sacramento Kings   
22           22          SAS    San Antonio       San Antonio Spurs   
23           23          OKC  Oklahoma City   Oklahoma City Thunder   
24           24          TOR        Toronto         Toronto Raptors   
25           25          UTA           Utah               Utah Jazz   
26           26          MEM        Memphis       Memphis Grizzlies   
27           27          WAS     Washington      Washington Wizards   
28           28          DET        Detroit         Detroit Pistons   
29           29          CHA      Charlotte       Charlotte Hornets   

       TEAM_ID       NICKNAME                 STATE  YEAR_FOUNDED  
0   1610612737          Hawks               Atlanta          1949  
1   1610612738        Celtics         Massachusetts          1946  
2   1610612739      Cavaliers                  Ohio          1970  
3   1610612740       Pelicans             Louisiana          2002  
4   1610612741          Bulls              Illinois          1966  
5   1610612742      Mavericks                 Texas          1980  
6   1610612743        Nuggets              Colorado          1976  
7   1610612744       Warriors            California          1946  
8   1610612745        Rockets                 Texas          1967  
9   1610612746       Clippers            California          1970  
10  1610612747         Lakers            California          1948  
11  1610612748           Heat               Florida          1988  
12  1610612749          Bucks             Wisconsin          1968  
13  1610612750   Timberwolves             Minnesota          1989  
14  1610612751           Nets              New York          1976  
15  1610612752         Knicks              New York          1946  
16  1610612753          Magic               Florida          1989  
17  1610612754         Pacers               Indiana          1976  
18  1610612755          76ers          Pennsylvania          1949  
19  1610612756           Suns               Arizona          1968  
20  1610612757  Trail Blazers                Oregon          1970  
21  1610612758          Kings            California          1948  
22  1610612759          Spurs                 Texas          1976  
23  1610612760        Thunder              Oklahoma          1967  
24  1610612761        Raptors               Ontario          1995  
25  1610612762           Jazz     

## Self Test - 1
- Rename "FULL_NAME" to "TEAM_NAME" 

In [8]:
#Your Code Here
NBA_Teams.rename(columns = {'FULL_NAME':'TEAM_NAME'}, inplace = True)
NBA_Teams

TEAM_NUMBER ABBREVIATION           CITY               TEAM_NAME  \
0             0          ATL        Atlanta           Atlanta Hawks   
1             1          BOS         Boston          Boston Celtics   
2             2          CLE      Cleveland     Cleveland Cavaliers   
3             3          NOP    New Orleans    New Orleans Pelicans   
4             4          CHI        Chicago           Chicago Bulls   
5             5          DAL         Dallas        Dallas Mavericks   
6             6          DEN         Denver          Denver Nuggets   
7             7          GSW   Golden State   Golden State Warriors   
8             8          HOU        Houston         Houston Rockets   
9             9          LAC    Los Angeles    Los Angeles Clippers   
10           10          LAL    Los Angeles      Los Angeles Lakers   
11           11          MIA          Miami              Miami Heat   
12           12          MIL      Milwaukee         Milwaukee Bucks   
13           13          MIN      Minnesota  Minnesota Timberwolves   
14           14          BKN       Brooklyn           Brooklyn Nets   
15           15          NYK       New York         New York Knicks   
16           16          ORL        Orlando           Orlando Magic   
17           17          IND        Indiana          Indiana Pacers   
18           18          PHI   Philadelphia      Philadelphia 76ers   
19           19          PHX        Phoenix            Phoenix Suns   
20           20          POR       Portland  Portland Trail Blazers   
21           21          SAC     Sacramento        Sacramento Kings   
22           22          SAS    San Antonio       San Antonio Spurs   
23           23          OKC  Oklahoma City   Oklahoma City Thunder   
24           24          TOR        Toronto         Toronto Raptors   
25           25          UTA           Utah               Utah Jazz   
26           26          MEM        Memphis       Memphis Grizzlies   
27           27          WAS     Washington      Washington Wizards   
28           28          DET        Detroit         Detroit Pistons   
29           29          CHA      Charlotte       Charlotte Hornets   

       TEAM_ID       NICKNAME                 STATE  YEAR_FOUNDED  
0   1610612737          Hawks               Atlanta          1949  
1   1610612738        Celtics         Massachusetts          1946  
2   1610612739      Cavaliers                  Ohio          1970  
3   1610612740       Pelicans             Louisiana          2002  
4   1610612741          Bulls              Illinois          1966  
5   1610612742      Mavericks                 Texas          1980  
6   1610612743        Nuggets              Colorado          1976  
7   1610612744       Warriors            California          1946  
8   1610612745        Rockets                 Texas          1967  
9   1610612746       Clippers            California          1970  
10  1610612747         Lakers            California          1948  
11  1610612748           Heat               Florida          1988  
12  1610612749          Bucks             Wisconsin          1968  
13  1610612750   Timberwolves             Minnesota          1989  
14  1610612751           Nets              New York          1976  
15  1610612752         Knicks              New York          1946  
16  1610612753          Magic               Florida          1989  
17  1610612754         Pacers               Indiana          1976  
18  1610612755          76ers          Pennsylvania          1949  
19  1610612756           Suns               Arizona          1968  
20  1610612757  Trail Blazers                Oregon          1970  
21  1610612758          Kings            California          1948  
22  1610612759          Spurs                 Texas          1976  
23  1610612760        Thunder              Oklahoma          1967  
24  1610612761        Raptors               Ontario          1995  
25  1610612762           Jazz     

### Dropping Variables (columns)

#### To drop a variable, i.e., to delete a column, we can use the “drop” command. 
- We need to provide the name of the variable;
- We also need to use the argument “axis=1” which tells Python that we are dropping a column, not a row. 

The variable "TEAM_NUMBER" has little meaning, let's drop it.

In [10]:
NBA_Teams.drop(['TEAM_NUMBER'], axis = 1, inplace = True)
NBA_Teams

ABBREVIATION           CITY               TEAM_NAME     TEAM_ID  \
0           ATL        Atlanta           Atlanta Hawks  1610612737   
1           BOS         Boston          Boston Celtics  1610612738   
2           CLE      Cleveland     Cleveland Cavaliers  1610612739   
3           NOP    New Orleans    New Orleans Pelicans  1610612740   
4           CHI        Chicago           Chicago Bulls  1610612741   
5           DAL         Dallas        Dallas Mavericks  1610612742   
6           DEN         Denver          Denver Nuggets  1610612743   
7           GSW   Golden State   Golden State Warriors  1610612744   
8           HOU        Houston         Houston Rockets  1610612745   
9           LAC    Los Angeles    Los Angeles Clippers  1610612746   
10          LAL    Los Angeles      Los Angeles Lakers  1610612747   
11          MIA          Miami              Miami Heat  1610612748   
12          MIL      Milwaukee         Milwaukee Bucks  1610612749   
13          MIN      Minnesota  Minnesota Timberwolves  1610612750   
14          BKN       Brooklyn           Brooklyn Nets  1610612751   
15          NYK       New York         New York Knicks  1610612752   
16          ORL        Orlando           Orlando Magic  1610612753   
17          IND        Indiana          Indiana Pacers  1610612754   
18          PHI   Philadelphia      Philadelphia 76ers  1610612755   
19          PHX        Phoenix            Phoenix Suns  1610612756   
20          POR       Portland  Portland Trail Blazers  1610612757   
21          SAC     Sacramento        Sacramento Kings  1610612758   
22          SAS    San Antonio       San Antonio Spurs  1610612759   
23          OKC  Oklahoma City   Oklahoma City Thunder  1610612760   
24          TOR        Toronto         Toronto Raptors  1610612761   
25          UTA           Utah               Utah Jazz  1610612762   
26          MEM        Memphis       Memphis Grizzlies  1610612763   
27          WAS     Washington      Washington Wizards  1610612764   
28          DET        Detroit         Detroit Pistons  1610612765   
29          CHA      Charlotte       Charlotte Hornets  1610612766   

         NICKNAME                 STATE  YEAR_FOUNDED  
0           Hawks               Atlanta          1949  
1         Celtics         Massachusetts          1946  
2       Cavaliers                  Ohio          1970  
3        Pelicans             Louisiana          2002  
4           Bulls              Illinois          1966  
5       Mavericks                 Texas          1980  
6         Nuggets              Colorado          1976  
7        Warriors            California          1946  
8         Rockets                 Texas          1967  
9        Clippers            California          1970  
10         Lakers            California          1948  
11           Heat               Florida          1988  
12          Bucks             Wisconsin          1968  
13   Timberwolves             Minnesota          1989  
14           Nets              New York          1976  
15         Knicks              New York          1946  
16          Magic               Florida          1989  
17         Pacers               Indiana          1976  
18          76ers          Pennsylvania          1949  
19           Suns               Arizona          1968  
20  Trail Blazers                Oregon          1970  
21          Kings            California          1948  
22          Spurs                 Texas          1976  
23        Thunder              Oklahoma          1967  
24        Raptors               Ontario          1995  
25           Jazz                  Utah          1974  
26      Grizzlies             Tennessee          1995  
27        Wizards  District of Columbia          1961  
28        Pistons              Michigan          1948  
29        Hornets        North Carolina          1988

#### Next we will work on game level data.

Import the game level dataset from our data repository. 
- We can display just first five rows of the dataset using the "head" command.


In [11]:
Games=pd.read_csv("../../Data/Week 2/basketball_games.csv")
Games.head()

SEASON_ID     TEAM_ID TEAM_ABBREVIATION           TEAM_NAME     GAME_ID  \
0      22019  1611661322               WAS  Washington Mystics  1021900079   
1      22019  1611661319               LVA      Las Vegas Aces  1021900079   
2      22019  1611661320               LAS  Los Angeles Sparks  1021900124   
3      22019  1611661323               CON     Connecticut Sun  1021900122   
4      22019  1611661317               PHO     Phoenix Mercury  1021900122   

  GAME_DATE      MATCHUP   WL  MIN  PTS  ...  FT_PCT  OREB  DREB  REB  AST  \
0  8/5/2019    WAS @ LVA  NaN  101   51  ...   0.750     3    10   13   14   
1  8/5/2019  LVA vs. WAS  NaN  100   36  ...   0.800     5    10   15   12   
2  8/1/2019  LAS vs. LVA    W  200   76  ...   1.000     5    28   33   19   
3  8/1/2019  CON vs. PHO    W  200   68  ...   0.786    12    28   40   18   
4  8/1/2019    PHO @ CON    L  199   62  ...   0.778     6    23   29   15   

   STL  BLK  TOV  PF  PLUS_MINUS  
0    2    4    5   7        12.2  
1    2    0    7   8       -11.8  
2    7    7   12  18         8.0  
3   13    0   16  15         6.0  
4    8    9   18  11        -6.0  

[5 rows x 28 columns]

_Upon importing the game data, we notice that the first five games are not NBA games, instead, they are WNBA games. Indeed, this dataset contains NBA games, WNBA games, NBA 2K (simulation video) games._ 

### Dropping observations (rows)

#### To drop an observation, we can use the index number on the left to specify the row we want to drop.
- The argument axis=0 specifies that we want to drop a row instead of a column.


In [12]:
Games.drop([0], axis=0, inplace=True)
Games.head()

SEASON_ID     TEAM_ID TEAM_ABBREVIATION           TEAM_NAME     GAME_ID  \
1      22019  1611661319               LVA      Las Vegas Aces  1021900079   
2      22019  1611661320               LAS  Los Angeles Sparks  1021900124   
3      22019  1611661323               CON     Connecticut Sun  1021900122   
4      22019  1611661317               PHO     Phoenix Mercury  1021900122   
5      22019  1611661319               LVA      Las Vegas Aces  1021900124   

  GAME_DATE      MATCHUP   WL  MIN  PTS  ...  FT_PCT  OREB  DREB  REB  AST  \
1  8/5/2019  LVA vs. WAS  NaN  100   36  ...   0.800     5    10   15   12   
2  8/1/2019  LAS vs. LVA    W  200   76  ...   1.000     5    28   33   19   
3  8/1/2019  CON vs. PHO    W  200   68  ...   0.786    12    28   40   18   
4  8/1/2019    PHO @ CON    L  199   62  ...   0.778     6    23   29   15   
5  8/1/2019    LVA @ LAS    L  200   68  ...   0.765    10    29   39   17   

   STL  BLK  TOV  PF  PLUS_MINUS  
1    2    0    7   8       -11.8  
2    7    7   12  18         8.0  
3   13    0   16  15         6.0  
4    8    9   18  11        -6.0  
5    7    4   14  13        -8.0  

[5 rows x 28 columns]

#### More often, we will drop observations based on certain conditions. 

For example, Las Vegas Aces is a women’s basketball team. If we are only going to focus on men’s basketball games, we will drop all the games played by Las Vegas Aces. In this case, we don’t have to use the “drop” function. We can specify our TEAM_NAME variables to be not equal to “Las Vegas Aces.”


In [13]:
Games = Games[Games.TEAM_NAME != 'Las Vegas Aces']
Games.head()

SEASON_ID     TEAM_ID TEAM_ABBREVIATION           TEAM_NAME     GAME_ID  \
2      22019  1611661320               LAS  Los Angeles Sparks  1021900124   
3      22019  1611661323               CON     Connecticut Sun  1021900122   
4      22019  1611661317               PHO     Phoenix Mercury  1021900122   
6      22019  1611661313               NYL    New York Liberty  1021900123   
7      22019  1611661321               DAL        Dallas Wings  1021900123   

  GAME_DATE      MATCHUP WL  MIN  PTS  ...  FT_PCT  OREB  DREB  REB  AST  STL  \
2  8/1/2019  LAS vs. LVA  W  200   76  ...   1.000     5    28   33   19    7   
3  8/1/2019  CON vs. PHO  W  200   68  ...   0.786    12    28   40   18   13   
4  8/1/2019    PHO @ CON  L  199   62  ...   0.778     6    23   29   15    8   
6  8/1/2019    NYL @ DAL  L  200   64  ...   0.800     6    21   27   15   10   
7  8/1/2019  DAL vs. NYL  W  200   87  ...   0.833     9    24   33   24   13   

   BLK  TOV  PF  PLUS_MINUS  
2    7   12  18         8.0  
3    0   16  15         6.0  
4    9   18  11        -6.0  
6    2   21  26       -23.0  
7    6   18  21        23.0  

[5 rows x 28 columns]

## Self Test - 2
- Drop all the Phoenix Mercury games

In [14]:
#Your Code Here
Games = Games[Games.TEAM_NAME != 'Phoenix Mercury']
Games

SEASON_ID     TEAM_ID TEAM_ABBREVIATION                  TEAM_NAME  \
2          22019  1611661320               LAS         Los Angeles Sparks   
3          22019  1611661323               CON            Connecticut Sun   
6          22019  1611661313               NYL           New York Liberty   
7          22019  1611661321               DAL               Dallas Wings   
8          22019  1611661325               IND              Indiana Fever   
9          22019  1611661330               ATL              Atlanta Dream   
11         22019  1611661322               WAS         Washington Mystics   
12         22019  1611661323               CON            Connecticut Sun   
13         22019  1611661321               DAL               Dallas Wings   
14         22019  1611661329               CHI                Chicago Sky   
16         32019  1611665409               EDD           Team Delle Donne   
17         32019  1611665410               WIL                Team Wilson   
18         42019         200               SGC                   76ers GC   
19         42019         205               CCG   Celtics Crossover Gaming   
20         42019         227               TWG            T-Wolves Gaming   
21         42019         227               TWG            T-Wolves Gaming   
22         42019         229               WGS      Warriors Gaming Squad   
23         42019         205               CCG   Celtics Crossover Gaming   
24         42019         229               WGS      Warriors Gaming Squad   
25         42019         200               SGC                   76ers GC   
26         42019         200               SGC                   76ers GC   
27         42019         227               TWG            T-Wolves Gaming   
28         42019         200               SGC                   76ers GC   
29         42019         213               KGG         Kings Guard Gaming   
30         42019         213               KGG         Kings Guard Gaming   
31         42019         220               PCG              Pacers Gaming   
32         42019         220               PCG              Pacers Gaming   
33         42019         227               TWG            T-Wolves Gaming   
34         42019         205               CCG   Celtics Crossover Gaming   
35         22019  1611661323               CON            Connecticut Sun   
...          ...         ...               ...                        ...   
29970      22012  1610612756               PHX               Phoenix Suns   
29971      22012  1610612752               NYK            New York Knicks   
29972      22012  1610612740               NOH        New Orleans Hornets   
29973      22012  1610612763               MEM          Memphis Grizzlies   
29974      22012  1610612743               DEN             Denver Nuggets   
29975      22012  1610612749               MIL            Milwaukee Bucks   
29976      22012  1610612744               GSW      Golden State Warriors   
29977      22012  1612709913               ERI              Erie BayHawks   
29978      22012  1610612750               MIN     Minnesota Timberwolves   
29979      22012  1610612766               CHA          Charlotte Bobcats   
29980      22012  1612709900               BAK            Bakersfield Jam   
29981      22012  1610612751               BKN              Brooklyn Nets   
29982      22012  1610612745               HOU            Houston Rockets   
29983      22012  1612709903               IDA             Idaho Stampede   
29984      22012  1612709902               SCW        Santa Cruz Warriors   
29985      22012  1610612764               WAS         Washington Wizards   
29986      22012  1612709914               RNO              Reno Bighorns   
29987      22012  1610612737               ATL              Atlanta Hawks   
29988      22012  1612709890               AUS               Austin Toros   
29989      22012  1612709904               SXF       Sioux Falls Skyforce   
299

## Merging Dataframes

#### We will only focus on NBA games. We could merge the NBA_Teams and Games datasets to filter out NBA games.

#### Teams are identified by the TEAM_ID. So, let’s merge the datasets by TEAM_ID. Since the variable “TEAM_NAME” is also present in both datasets, we could also include this variable as a criteria to merge the datasets so that in our new dataset, there is no duplicate variables.

In [15]:
NBA_Games = pd.merge(NBA_Teams, Games, on = ['TEAM_ID', 'TEAM_NAME'])
NBA_Games

ABBREVIATION       CITY          TEAM_NAME     TEAM_ID NICKNAME  \
0              ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
1              ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
2              ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
3              ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
4              ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
5              ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
6              ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
7              ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
8              ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
9              ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
10             ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
11             ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
12             ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
13             ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
14             ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
15             ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
16             ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
17             ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
18             ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
19             ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
20             ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
21             ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
22             ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
23             ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
24             ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
25             ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
26             ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
27             ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
28             ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
29             ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
...            ...        ...                ...         ...      ...   
18391          CHA  Charlotte  Charlotte Hornets  1610612766  Hornets   
18392          CHA  Charlotte  Charlotte Hornets  1610612766  Hornets   
18393          CHA  Charlotte  Charlotte Hornets  1610612766  Hornets   
18394          CHA  Charlotte  Charlotte Hornets  1610612766  Hornets   
18395          CHA  Charlotte  Charlotte Hornets  1610612766  Hornets   
18396          CHA  Charlotte  Charlotte Hornets  1610612766  Hornets   
18397          CHA  Charlotte  Charlotte Hornets  1610612766  Hornets   
18398          CHA  Charlotte  Charlotte Hornets  1610612766  Hornets   
18399          CHA  Charlotte  Charlotte Hornets  1610612766  Hornets   
18400          CHA  Charlotte  Charlotte Hornets  1610612766  Hornets   
18401          CHA  Charlotte  Charlotte Hornets  1610612766  Hornets   
18402          CHA  Charlotte  Charlotte Hornets  1610612766  Hornets   
18403          CHA  Charlotte  Charlotte Hornets  1610612766  Hornets   
18404          CHA  Charlotte  Charlotte Hornets  1610612766  Hornets   
18405          CHA  Charlotte  Charlotte Hornets  1610612766  Hornets   
18406          CHA  Charlotte  Charlotte Hornets  1610612766  Hornets   
18407          CHA  Charlotte  Charlotte Hornets  1610612766  Hornets   
18408          CHA  Charlotte  Charlotte Hornets  1610612766  Hornets   
18409          CHA  Charlotte  Charlotte Hornets  1610612766  Hornets   
18410          CHA  Charlotte  Charlotte Hornets  1610612766  Hornets   
18411          CHA  Charlotte  Charlotte Hornets  1610612766  Hornets   
18412          CHA  Charlotte  Charlotte Hornets  1610612766  Hornets   
18413          CHA  Charlotte  Charlotte Hornets  1610612766  Ho

### Understanding and cleaning the merged dataset

_As you can tell, the merged dataset has a lot more variables and Python cannot fit all of them in the screen._ 

#### We can obtain the list of variables using the “columns” command. This provides us a full list of variables in our dataset.


In [16]:
NBA_Games.columns

Index(['ABBREVIATION', 'CITY', 'TEAM_NAME', 'TEAM_ID', 'NICKNAME', 'STATE',
       'YEAR_FOUNDED', 'SEASON_ID', 'TEAM_ABBREVIATION', 'GAME_ID',
       'GAME_DATE', 'MATCHUP', 'WL', 'MIN', 'PTS', 'FGM', 'FGA', 'FG_PCT',
       'FG3M', 'FG3A', 'FG3_PCT', 'FTM', 'FTA', 'FT_PCT', 'OREB', 'DREB',
       'REB', 'AST', 'STL', 'BLK', 'TOV', 'PF', 'PLUS_MINUS'],
      dtype='object')

#### Data Cleaning
The variable “ABBREVIATION” AND “TEAM_ABBREVIATION” carry the same information and it is not necessary to keep both of them. 
- Delete "ABBREVIATION" 


In [17]:
NBA_Games.drop(['ABBREVIATION'], axis = 1, inplace = True, errors = 'ignore')

## Self Test - 3
- Find the number of observations and the number of variables in the dataset

In [18]:
#Your Code Here
NBA_Games.shape

(18421, 32)

The merged dataset is sorted by the criteria we use to merge the datasets. Thus, the NBA_Games dataset is currently sorted by "TEAM_ID." We may be interested to sort the data by other criteria, for example, the date of the game. 

#### We can do so by using the “sort_values” option. 

In our dataset, "GAME_ID" is created based on the date of the game. We can sort the games by “GAME_ID” and display the 20 most recent games.


In [19]:
NBA_Games.sort_values(by=['GAME_ID'], ascending=[False]).head(20)

CITY            TEAM_NAME     TEAM_ID   NICKNAME      STATE  \
13449  San Antonio    San Antonio Spurs  1610612759      Spurs      Texas   
15445         Utah            Utah Jazz  1610612762       Jazz       Utah   
1304     Cleveland  Cleveland Cavaliers  1610612739  Cavaliers       Ohio   
16076      Memphis    Memphis Grizzlies  1610612763  Grizzlies  Tennessee   
15446         Utah            Utah Jazz  1610612762       Jazz       Utah   
1305     Cleveland  Cleveland Cavaliers  1610612739  Cavaliers       Ohio   
13450  San Antonio    San Antonio Spurs  1610612759      Spurs      Texas   
16077      Memphis    Memphis Grizzlies  1610612763  Grizzlies  Tennessee   
15447         Utah            Utah Jazz  1610612762       Jazz       Utah   
16078      Memphis    Memphis Grizzlies  1610612763  Grizzlies  Tennessee   
1306     Cleveland  Cleveland Cavaliers  1610612739  Cavaliers       Ohio   
13451  San Antonio    San Antonio Spurs  1610612759      Spurs      Texas   
15545         Utah            Utah Jazz  1610612762       Jazz       Utah   
99         Atlanta        Atlanta Hawks  1610612737      Hawks    Atlanta   
16173      Memphis    Memphis Grizzlies  1610612763  Grizzlies  Tennessee   
13551  San Antonio    San Antonio Spurs  1610612759      Spurs      Texas   
15546         Utah            Utah Jazz  1610612762       Jazz       Utah   
16174      Memphis    Memphis Grizzlies  1610612763  Grizzlies  Tennessee   
13552  San Antonio    San Antonio Spurs  1610612759      Spurs      Texas   
100        Atlanta        Atlanta Hawks  1610612737      Hawks    Atlanta   

       YEAR_FOUNDED  SEASON_ID TEAM_ABBREVIATION     GAME_ID GAME_DATE  ...  \
13449          1976      22019               SAS  1621900006  7/3/2019  ...   
15445          1974      22019               UTA  1621900006  7/3/2019  ...   
1304           1970      22019               CLE  1621900005  7/3/2019  ...   
16076          1995      22019               MEM  1621900005  7/3/2019  ...   
15446          1974      22019               UTA  1621900004  7/2/2019  ...   
1305           1970      22019               CLE  1621900004  7/2/2019  ...   
13450          1976      22019               SAS  1621900003  7/2/2019  ...   
16077          1995      22019               MEM  1621900003  7/2/2019  ...   
15447          1974      22019               UTA  1621900002  7/1/2019  ...   
16078          1995      22019               MEM  1621900002  7/1/2019  ...   
1306           1970      22019               CLE  1621900001  7/1/2019  ...   
13451          1976      22019               SAS  1621900001  7/1/2019  ...   
15545          1974      22018               UTA  1621800006  7/5/2018  ...   
99             1949      22018               ATL  1621800006  7/5/2018  ...   
16173          1995      22018               MEM  1621800005  7/5/2018  ...   
13551          1976      22018               SAS  1621800005  7/5/2018  ...   
15546          1974      22018               UTA  1621800004  7/3/2018  ...   
16174          1995      22018               MEM  1621800004  7/3/2018  ...   
13552          1976      22018               SAS  1621800003  7/3/2018  ...   
100            1949      22018               ATL  1621800003  7/3/2018  ...   

      FT_PCT OREB  DREB  REB  AST  STL  BLK  TOV  PF  PLUS_MINUS  
13449  0.615    9    29   38   12    5    3   11  14        -7.2  
15445  0.684   13    31   44   15    6    5   15  25         5.8  
1304   0.737    3    26   29   15    9    3   17  12       -13.0  
16076  0.692   11    36   47   19    8    5   14  19        13.0  
15446  0.750   11    34   45   18    3    4   14  18        14.6  
1305   0.682    6    27   33   13    6    0    6  14       -15.0  
13450  0.870   10    35   45   24   10    7   18  24        16.6  
16077  0.867   10    29   39   17    6    5   17  18       -15.8  
15447  0.462   14    27   41   16   10    7   19  26        -6.6  
16078  0.704    8    32   40   14   14    4   20  19         7.4  

## Missing Values

### Before we move on to doing any data analyses, we usually need to check if there is any missing value, that is, the source may have failed to collect some information. 

#### We can use the info() command which will return the total number of observations that have real values. By looking at these total numbers, we can see if there is any variable with missing value.


In [20]:
NBA_Games.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18421 entries, 0 to 18420
Data columns (total 32 columns):
CITY                 18421 non-null object
TEAM_NAME            18421 non-null object
TEAM_ID              18421 non-null int64
NICKNAME             18421 non-null object
STATE                18421 non-null object
YEAR_FOUNDED         18421 non-null int64
SEASON_ID            18421 non-null int64
TEAM_ABBREVIATION    18421 non-null object
GAME_ID              18421 non-null int64
GAME_DATE            18421 non-null object
MATCHUP              18421 non-null object
WL                   18414 non-null object
MIN                  18421 non-null int64
PTS                  18421 non-null int64
FGM                  18421 non-null int64
FGA                  18421 non-null int64
FG_PCT               18419 non-null float64
FG3M                 18421 non-null int64
FG3A                 18421 non-null int64
FG3_PCT              18418 non-null float64
FTM                  18421 non-null int

_The total number of rows is 18421, so there are missing values in variable WL, FG_PCT, FG3_PCT, and FT_PCT._

#### Detecting missing values
We can use the isnull() function and the notnull() function to detect where the missing values are.


In [21]:
NBA_Games.notnull()

CITY  TEAM_NAME  TEAM_ID  NICKNAME  STATE  YEAR_FOUNDED  SEASON_ID  \
0      True       True     True      True   True          True       True   
1      True       True     True      True   True          True       True   
2      True       True     True      True   True          True       True   
3      True       True     True      True   True          True       True   
4      True       True     True      True   True          True       True   
5      True       True     True      True   True          True       True   
6      True       True     True      True   True          True       True   
7      True       True     True      True   True          True       True   
8      True       True     True      True   True          True       True   
9      True       True     True      True   True          True       True   
10     True       True     True      True   True          True       True   
11     True       True     True      True   True          True       True   
12     True       True     True      True   True          True       True   
13     True       True     True      True   True          True       True   
14     True       True     True      True   True          True       True   
15     True       True     True      True   True          True       True   
16     True       True     True      True   True          True       True   
17     True       True     True      True   True          True       True   
18     True       True     True      True   True          True       True   
19     True       True     True      True   True          True       True   
20     True       True     True      True   True          True       True   
21     True       True     True      True   True          True       True   
22     True       True     True      True   True          True       True   
23     True       True     True      True   True          True       True   
24     True       True     True      True   True          True       True   
25     True       True     True      True   True          True       True   
26     True       True     True      True   True          True       True   
27     True       True     True      True   True          True       True   
28     True       True     True      True   True          True       True   
29     True       True     True      True   True          True       True   
...     ...        ...      ...       ...    ...           ...        ...   
18391  True       True     True      True   True          True       True   
18392  True       True     True      True   True          True       True   
18393  True       True     True      True   True          True       True   
18394  True       True     True      True   True          True       True   
18395  True       True     True      True   True          True       True   
18396  True       True     True      True   True          True       True   
18397  True       True     True      True   True          True       True   
18398  True       True     True      True   True          True       True   
18399  True       True     True      True   True          True       True   
18400  True       True     True      True   True          True       True   
18401  True       True     True      True   True          True       True   
18402  True       True     True      True   True          True       True   
18403  True       True     True      True   True          True       True   
18404  True       True     True      True   True          True       True   
18405  True       True     True      True   True          True       True   
18406  True       True     True      True   True          True       True   
18407  True       True     True      True   True          True       True   
18408  True       True     True      True   True          True       True   
18409  True       True     True      True   True          True       True   
18410  True       True     True      True   True          True       True   
184

### Handling Missing Values

There are two main approaches to handle missing values. 
- First, we can simply drop the observations with missing value.

#### Drop observations with missing value in the variable "FG_PCT"

In [22]:
NBA_Games=NBA_Games[pd.notnull(NBA_Games["FG_PCT"])]
NBA_Games.shape

(18419, 32)

- Second, we can replace the missing values with valid values (Imputation), such as mean and median.

#### We can use the fillna() command to replace missing values with the mean or the median of the variable.


In [23]:
NBA_Games=NBA_Games.fillna(NBA_Games.mean())
NBA_Games.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18419 entries, 0 to 18420
Data columns (total 32 columns):
CITY                 18419 non-null object
TEAM_NAME            18419 non-null object
TEAM_ID              18419 non-null int64
NICKNAME             18419 non-null object
STATE                18419 non-null object
YEAR_FOUNDED         18419 non-null int64
SEASON_ID            18419 non-null int64
TEAM_ABBREVIATION    18419 non-null object
GAME_ID              18419 non-null int64
GAME_DATE            18419 non-null object
MATCHUP              18419 non-null object
WL                   18414 non-null object
MIN                  18419 non-null int64
PTS                  18419 non-null int64
FGM                  18419 non-null int64
FGA                  18419 non-null int64
FG_PCT               18419 non-null float64
FG3M                 18419 non-null int64
FG3A                 18419 non-null int64
FG3_PCT              18419 non-null float64
FTM                  18419 non-null int

## Creating variables
We can create a variable equals to the total number of goals made.


In [24]:
NBA_Games['GM']=NBA_Games['FGM']+NBA_Games['FG3M']+NBA_Games['FTM']

## Self Test - 4
- Create a variable called "GA" equals to the total number of goals attempted.

In [27]:
Games.columns

Index(['SEASON_ID', 'TEAM_ID', 'TEAM_ABBREVIATION', 'TEAM_NAME', 'GAME_ID',
       'GAME_DATE', 'MATCHUP', 'WL', 'MIN', 'PTS', 'FGM', 'FGA', 'FG_PCT',
       'FG3M', 'FG3A', 'FG3_PCT', 'FTM', 'FTA', 'FT_PCT', 'OREB', 'DREB',
       'REB', 'AST', 'STL', 'BLK', 'TOV', 'PF', 'PLUS_MINUS'],
      dtype='object')

In [28]:
#Your Code Here
NBA_Games['GA'] = NBA_Games['FTA'] + NBA_Games['FG3A'] + NBA_Games['FGA']

### Create variables based on conditions
- We can create a variable conditional on the value of another variable.

For example, we can create a variable "RESULT" that equals to 1 if the team won the game and 0 otherwise. The result of the game can be captured in the points of the team receive, whether it was positive or negative.


In [29]:
NBA_Games['RESULT'] = np.where(NBA_Games['PLUS_MINUS']>0, 'W', 'L')

We will now drop this newly created "RESULT" variable.


In [30]:
NBA_Games.drop(['RESULT'], axis=1, inplace=True)

### Create a variable within group
In the dataset, each game has two observations, one represents the statistics of the home team, one represents those of the away team. Both observations have the same GAME_ID. We can create a variable "POINT_DIFF" that equals the difference between the points earned by the two teams.

We will first sort the data not only by the "GAME_ID" but also by the result "WL".


In [31]:
NBA_Games.sort_values(['GAME_ID','WL'], inplace=True)
NBA_Games["POINT_DIFF"]=NBA_Games.groupby(["GAME_ID"])["PTS"].diff()

The "POINT_DIFF" variable only has the point difference for the winning team, we need to impute the point difference for the losing team as well.



In [32]:
NBA_Games['POINT_DIFF'] = NBA_Games['POINT_DIFF'].fillna(NBA_Games.groupby('GAME_ID')['POINT_DIFF'].transform('mean'))

- We can also drop all observations with missing value in at least one variable using the "dropna()" command.


In [33]:
NBA_Games=NBA_Games.dropna()
NBA_Games.shape

(17779, 35)

### Creating new dataframe

#### Create a new dataframe that aggregates information by group

Sometimes we may want to work with season level data rather than team level data. We can create a new dataset that includes aggregate information of team statistics in each season.




In [34]:
NBA_Team_Stats=NBA_Games.groupby(['TEAM_ID', 'SEASON_ID'])['PTS','FGM','FGA','FG_PCT','FG3M','FG3A','FG3_PCT','FTM','FTA','FT_PCT','OREB','DREB','REB','AST','STL','BLK','TOV','PF','PLUS_MINUS'].sum()
display(NBA_Team_Stats)

PTS   FGM   FGA  FG_PCT  FG3M  FG3A  FG3_PCT   FTM  \
TEAM_ID    SEASON_ID                                                        
1610612737 12013       523   193   459   2.520    34   131    1.576   103   
           12014       708   247   554   3.122    66   182    2.549   148   
           12015       652   226   548   2.896    59   176    2.340   141   
           12016       686   261   593   3.083    50   159    2.140   114   
           12017       480   167   410   2.045    51   156    1.611    95   
           12018       563   206   445   2.312    63   189    1.659    88   
           22012      2074   795  1701   9.824   155   454    7.206   329   
           22013      8272  3061  6719  37.846   747  2085   29.419  1403   
           22014      8786  3253  7017  40.399   848  2256   32.719  1432   
           22015      8701  3265  7163  39.250   833  2403   29.818  1338   
           22016      8736  3219  7142  38.784   760  2214   29.491  1538   
           22017      8683  3206  7208  37.909   919  2567   30.388  1352   
           22018      9742  3555  7987  39.210  1121  3226   30.483  1511   
           22019       381   127   342   1.873    43   136    1.571    84   
           42012       536   196   464   2.533    37   112    1.900   107   
           42013       647   212   551   2.729    79   230    2.483   144   
           42014      1568   590  1369   6.907   145   458    5.030   243   
           42015      1001   366   848   4.337   106   318    3.332   163   
           42016       633   225   507   2.662    50   164    1.798   133   
1610612738 12013       754   276   654   3.404    54   186    2.313   148   
           12014       788   278   652   3.415    82   230    2.866   150   
           12015       495   187   438   2.154    44   121    1.828    77   
           12016       735   271   595   3.210    71   207    2.409   122   
           12017       425   155   361   1.724    50   124    1.613    65   
           12018       409   148   370   1.608    34   138    1.012    79   
           22012      1679   639  1327   8.192   119   316    6.483   282   
           22013      8044  3050  6991  36.818   580  1772   27.093  1364   
           22014      8728  3335  7580  38.363   701  2173   28.094  1357   
           22015      9124  3357  7697  38.360   762  2292   29.406  1648   
           22016      9319  3315  7330  39.823  1022  2844   31.486  1667   
...                    ...   ...   ...     ...   ...   ...      ...   ...   
1610612764 42016      1405   521  1139   5.959   109   344    4.170   254   
           42017       643   238   511   2.793    48   131    2.239   119   
1610612765 12013       655   228   522   3.068    51   141    2.492   148   
           12014       698   249   570   3.067    59   171    2.399   141   
           12015       824   302   667   3.641    78   212    2.940   142   
           12016       565   216   485   2.680    40   129    1.804    93   
           12017       517   202   437   2.311    43   119    1.796    70   
           12018       524   190   467   2.042    67   204    1.628    77   
           22012      1534   585  1256   7.487    95   297    5.077   269   
           22013      8363  3194  7181  37.350   513  1625   25.813  1462   
           22014      8514  3187  7358  37.791   746  2163   29.631  1394   
           22015      8393  3121  7117  36.482   743  2150   28.588  1408   
           22016      8351  3270  7298  37.161   638  1948   26.954  1173   
           22017      8754  3298  7310  38.371   913  2438   31.727  1245   
           22018      9020  3271  7500  37.520  1004  2946   29.104  1474   
           22019       376   131   311   1.703    48   134    1.449    66   
           42015       380   144   301   1.916    35    93    1.430    57   
           42018       392   150   387   1.549    44   133    1.317    48   
1610612766 12014       759   278   680   3.260    53   170    2.444   150   
           12015   

Notice that the newly created dataset has two levels of index, the "TEAM_ID" and "SEASON_ID"

#### If we want to convert these two indexes back as variables, we can use the "reset_index" command.


In [35]:
NBA_Team_Stats=NBA_Team_Stats.reset_index()
display(NBA_Team_Stats)

TEAM_ID  SEASON_ID   PTS   FGM   FGA  FG_PCT  FG3M  FG3A  FG3_PCT  \
0    1610612737      12013   523   193   459   2.520    34   131    1.576   
1    1610612737      12014   708   247   554   3.122    66   182    2.549   
2    1610612737      12015   652   226   548   2.896    59   176    2.340   
3    1610612737      12016   686   261   593   3.083    50   159    2.140   
4    1610612737      12017   480   167   410   2.045    51   156    1.611   
5    1610612737      12018   563   206   445   2.312    63   189    1.659   
6    1610612737      22012  2074   795  1701   9.824   155   454    7.206   
7    1610612737      22013  8272  3061  6719  37.846   747  2085   29.419   
8    1610612737      22014  8786  3253  7017  40.399   848  2256   32.719   
9    1610612737      22015  8701  3265  7163  39.250   833  2403   29.818   
10   1610612737      22016  8736  3219  7142  38.784   760  2214   29.491   
11   1610612737      22017  8683  3206  7208  37.909   919  2567   30.388   
12   1610612737      22018  9742  3555  7987  39.210  1121  3226   30.483   
13   1610612737      22019   381   127   342   1.873    43   136    1.571   
14   1610612737      42012   536   196   464   2.533    37   112    1.900   
15   1610612737      42013   647   212   551   2.729    79   230    2.483   
16   1610612737      42014  1568   590  1369   6.907   145   458    5.030   
17   1610612737      42015  1001   366   848   4.337   106   318    3.332   
18   1610612737      42016   633   225   507   2.662    50   164    1.798   
19   1610612738      12013   754   276   654   3.404    54   186    2.313   
20   1610612738      12014   788   278   652   3.415    82   230    2.866   
21   1610612738      12015   495   187   438   2.154    44   121    1.828   
22   1610612738      12016   735   271   595   3.210    71   207    2.409   
23   1610612738      12017   425   155   361   1.724    50   124    1.613   
24   1610612738      12018   409   148   370   1.608    34   138    1.012   
25   1610612738      22012  1679   639  1327   8.192   119   316    6.483   
26   1610612738      22013  8044  3050  6991  36.818   580  1772   27.093   
27   1610612738      22014  8728  3335  7580  38.363   701  2173   28.094   
28   1610612738      22015  9124  3357  7697  38.360   762  2292   29.406   
29   1610612738      22016  9319  3315  7330  39.823  1022  2844   31.486   
..          ...        ...   ...   ...   ...     ...   ...   ...      ...   
485  1610612764      42016  1405   521  1139   5.959   109   344    4.170   
486  1610612764      42017   643   238   511   2.793    48   131    2.239   
487  1610612765      12013   655   228   522   3.068    51   141    2.492   
488  1610612765      12014   698   249   570   3.067    59   171    2.399   
489  1610612765      12015   824   302   667   3.641    78   212    2.940   
490  1610612765      12016   565   216   485   2.680    40   129    1.804   
491  1610612765      12017   517   202   437   2.311    43   119    1.796   
492  1610612765      12018   524   190   467   2.042    67   204    1.628   
493  1610612765      22012  1534   585  1256   7.487    95   297    5.077   
494  1610612765      22013  8363  3194  7181  37.350   513  1625   25.813   
495  1610612765      22014  8514  3187  7358  37.791   746  2163   29.631   
496  1610612765      22015  8393  3121  7117  36.482   743  2150   28.588   
497  1610612765      22016  8351  3270  7298  37.161   638  1948   26.954   
498  1610612765      22017  8754  3298  7310  38.371   913  2438   31.727   
499  1610612765      22018  9020  3271  7500  37.520  1004  2946   29.104   
500  1610612765      22019   376   131   311   1.703    48   134    1.449   
501  1610612765      42015   380   144   301   1.916    35    93    1.430   
502  1610612765      42018   392   150   387   1.549    44   133    1.317   
503  1610612766      12014   759   278   680   3.260    53   170    2.444   
504  1610612766      12015   572   194   476   2.446    55   158    2.066   
505 

### We can create a variable that equals to the total number of observations within a specified group using the size() command.
- Create a variable that equals to the total number of games played by a team in each season, name this variable "GAME_COUNT".


In [36]:
NBA_Game_Count=NBA_Games.groupby(['TEAM_ID','SEASON_ID']).size().reset_index(name='GAME_COUNT')
display(NBA_Game_Count)

TEAM_ID  SEASON_ID  GAME_COUNT
0    1610612737      12013           6
1    1610612737      12014           7
2    1610612737      12015           7
3    1610612737      12016           7
4    1610612737      12017           5
5    1610612737      12018           5
6    1610612737      22012          21
7    1610612737      22013          83
8    1610612737      22014          87
9    1610612737      22015          86
10   1610612737      22016          86
11   1610612737      22017          85
12   1610612737      22018          88
13   1610612737      22019           5
14   1610612737      42012           6
15   1610612737      42013           7
16   1610612737      42014          16
17   1610612737      42015          10
18   1610612737      42016           6
19   1610612738      12013           8
20   1610612738      12014           8
21   1610612738      12015           5
22   1610612738      12016           7
23   1610612738      12017           4
24   1610612738      12018           4
25   1610612738      22012          17
26   1610612738      22013          84
27   1610612738      22014          87
28   1610612738      22015          88
29   1610612738      22016          88
..          ...        ...         ...
485  1610612764      42016          13
486  1610612764      42017           6
487  1610612765      12013           7
488  1610612765      12014           7
489  1610612765      12015           8
490  1610612765      12016           6
491  1610612765      12017           5
492  1610612765      12018           5
493  1610612765      22012          16
494  1610612765      22013          84
495  1610612765      22014          87
496  1610612765      22015          83
497  1610612765      22016          83
498  1610612765      22017          85
499  1610612765      22018          86
500  1610612765      22019           4
501  1610612765      42015           4
502  1610612765      42018           4
503  1610612766      12014           8
504  1610612766      12015           6
505  1610612766      12016           7
506  1610612766      12017           5
507  1610612766      12018           5
508  1610612766      22014          89
509  1610612766      22015          84
510  1610612766      22016          84
511  1610612766      22017          85
512  1610612766      22018          85
513  1610612766      22019           4
514  1610612766      42015           7

[515 rows x 3 columns]

## Saving data

### We can save a dataframe by exporting the edited dataframe to csv file using the “to_csv” command.
- Save merged data as a csv file 
_We can use the "index=False" command to save the data without adding the index as a column in the csv file_

In [ ]:
NBA_Games.to_csv("../../Data/Week 2/NBA_Games.csv", index=False)